docker-compose.yaml

```diff
@@ -0,0 +1,12 @@
+version: "3.3"
+services:
+  waterfall:
+    container_name: waterfall
+    image: ghcr.io/matrix-org/waterfall:main
+    network_mode: host
+    restart: always
+    environment:
```

Comment:
Is there a way to avoid using env vars? We've just started transitioning from using env vars to using a config file, so I think we'd prefer not to use env vars in general

Reply:
Good question, let me elaborate on that 🙂 

So basically it all comes down to the flexibility of the setup so that we remain compatible with different ways to deploy and run the server as an image. Environment variables do not seem like the best idea when testing things locally or running the server in a bare Linux environment (which is the reason why we don't really obsolete the usage of the config file, it's still an option), however, it does not necessarily play well when we talk about automated deployments with e.g. k8s.

The problem with the config file is that it poses the question: how do we pass the config file to the image? Essentially we could do either of these:

- Embed the config file directly into the built image, i.e. making it part of the image. This is a bit problematic as it means that we must preconfigure the server at the moment the image is built which makes the image pretty much useless for any use case outside of the environment, the config of which was bundled together with the image.
- Another possibility is to supply the config file together with the image, but it means we would need to deliver an additional file with our `docker-compose` and mount the file to be accessible to the container on a runtime (while making sure that permissions, etc are all set right), in addition, we would need to pass the file name of the config explicitly as part of the `docker-compose.yml` which also makes the compose file a bit less useful.

This is normally solved by just passing the environment variables to the image as it also makes it easy to configure them per environment, i.e. we just deploy a single `docker-compose.yml` and we might provide a `.env` file that is automatically read when the container is started, meaning that the `docker-compose.yml` is environment-agnostic (could be used for both development and production environments) and all the configuration is just a part of the `.env` file as simple pairs of key=value.

There is an issue with the way I implemented which I realized upon reading your comment: since I introduced a separate variable for each configuration, it means that any time we add an additional configuration, we must both update the format of the config file **and** add handling of an additional environment variable to the function that reads environment variables. This is not very nice (copy-paste and syncing logic in two places), so I'm thinking about eliminating it the same way LiveKit SFU implemented it! What they did is that their `docker-compose.yml` file has a variable called `LIVEKIT_CONFIG` which is simply just a YAML. I.e. instead of creating separate variables for all configs, they allowed passing a `LIVEKIT_CONFIG` variable which is a regular body of the config file.

Perhaps I need to update it to have the same behavior to remain flexible and lower our maintenance effort!

Reply:
@SimonBrandner, updated the PR to address all your points, ptal 🙂